document/core/valid/instructions.rst

```diff
@@ -154,6 +154,263 @@ Numeric Instructions
    }
 
 
+.. index:: simd instruction
+   pair: validation; instruction
+   single: abstract syntax; instruction
+
+.. _valid-instr-simd:
+
+SIMD Instructions
+~~~~~~~~~~~~~~~~~
+
+.. _aux-unpacked:
+
+SIMD instructions can have a prefix to describe the :ref:`shape <syntax-simd-shape>` of the operand. Packed numeric types, |I8| and |I16|, are not :ref:`value type <syntax-valtype>`, we define an auxiliary function to map such packed types into value types:
+
+.. math::
+   \begin{array}{lll@{\qquad}l}
+   \unpacked(\K{i8x16}) &=& \I32 \\
+   \unpacked(\K{i16x8}) &=& \I32 \\
+   \unpacked(t\K{x}N) &=& t
+   \end{array}
+
+
+We also define an auxiliary function to get number of packed numeric types in a |V128|:
+
+.. _aux-lanes:
+
+.. math::
+   \begin{array}{lll@{\qquad}l}
+   \lanes(t\K{x}N) &=& N
+   \end{array}
+
+
+.. _valid-vconst:
+
+:math:`\V128\K{.}\VCONST~c`
+...........................
+
+* The instruction is valid with type :math:`[] \to [\V128]`.
+
+.. math::
+   \frac{
+   }{
+     C \vdashinstr \V128\K{.}\VCONST~c : [] \to [\V128]
+   }
+
+
+.. _valid-vsunop:
+
+:math:`\V128\K{.}\vsunop`
+.........................
+
+* The instruction is valid with type :math:`[\V128] \to [\V128]`.
+
+.. math::
+   \frac{
+   }{
+     C \vdashinstr \V128\K{.}\vsunop : [\V128] \to [\V128]
+   }
+
+
+.. _valid-vsbinop:
+
+:math:`\V128\K{.}\vsbinop`
+..........................
+
+* The instruction is valid with type :math:`[\V128~\V128] \to [\V128]`.
+
+.. math::
+   \frac{
+   }{
+     C \vdashinstr \V128\K{.}\vsbinop : [\V128~\V128] \to [\V128]
+   }
+
+
+.. _valid-vsternop:
+
+:math:`\V128\K{.}\vsternop`
+...........................
+
+* The instruction is valid with type :math:`[\V128~\V128~\V128] \to [\V128]`.
+
+.. math::
+   \frac{
+   }{
+     C \vdashinstr \V128\K{.}\vsternop : [\V128~\V128~\V128] \to [\V128]
+   }
+
+
+.. _valid-simd-shuffle:
+
+:math:`\K{v8x16.}\SHUFFLE~\laneidx^{16}`
+........................................
+
+* For all :math:`\laneidx_i`, in :math:`\laneidx^{16}`, :math:`\laneidx_i` must be smaller than :math:`32`.
+
+* The instruction is valid with type :math:`[\V128~\V128] \to [\V128]`.
+
+.. math::
+   \frac{
+     (\laneidx < 32)^{16}
```

Comment:
Why 32, shouldn't it be 16? (See also my suggestion to make laneidx an i4, which would make this condition obsolete.)

Reply:
For shuffle, indices can go to 31, since `[0,15]` select from the first operand, and `[16,31]` select from the second operand. (see https://github.com/WebAssembly/simd/blob/master/proposals/simd/SIMD.md#shuffling-using-immediate-indices)

Reply:
Because of this, I changed #346 to be a `u8` instead of `u4`. As defined in the semantics, `u8` lane indices are accepted (no decoding error), but will be rejected by validation.
(Text format might need some tweaking, since those rejects lane indices that are out of range at the parsing stage. I'll look at those separately)

Reply:
Hm, okay, I wonder if that's worth keeping that separate? It's a strange form of index that I'd argue isn't really a proper laneidx. OTOH, maybe that's not worth the trouble.

Reply:
Say we make `\laneidx ::= u4` in the syntax. In the text format, if we specify that lane indices must be `u4` as well, it means that values like 255 will not be accepted in the text format, which I guess is a parse error?

Right now the implementation will accept 255, and reject it during validation. Which requires that in the text format, lane indices be `u8`. But then we won't be able to create the syntax, since it's a `u4`.

Maybe I'm confused, but for the above reason, I think keeping them the same (`u8`) will be better - we can keep the text -> syntax as currently implemented, and instructions can reject out of bounds indices as necessary during validation.

Reply:
Well, yes, if we defined it to be u4, then larger values would have to be parse errors, and the reference implementation would have to be adjusted accordingly. I don't see a problem with that, but probably not worth it, at least for now.